### PR TITLE
Faster card creation

### PR DESCRIPTION
**Pull request by stexus**
Hi, I was using my fork of mpvacious for the past year without updates, but when I finally got around to updating it the conflict fixing was a bit of a headache, so I thought I'd see if you would be interested in merging this upstream. It is pretty hacky at the moment but it works. There would have to be some additions for config reading to change keybinds I think. I wanted to see what you thought before making the effort for that as I don't need it for my personal setup haha

I basically wanted a faster way to make basic cards where I don't manually set the start/end times. Normally the default Ctrl+M binding does the job; however, there are cases I want to either use multiple lines (usually 2) and/or update multiple notes at once, say when I encounter two unknown words within the same context or sentence. Mpvacious at the moment has incredible flexibility but it comes at the cost of a lot more keystrokes than I'm used to. 

What I added was a "quick card creation menu" where card creation for most subtitled works can be done in two or less keystrokes. It opens with 'g' (by default) and asks for a number. I can enter [0-9] to overwrite the last note with the number of subs chosen. alt+g would let me change the number of cards to update, followed by the number of subs. 

This relies on the fact that mpv has actually seen the next subtitle, but that will usually already be the case as the user wouldn't be trying to add lines they haven't seen yet to the card. 

I've successfully created plenty of cards with this now and I love having both the quick card creation as well as the advanced menu flexibility when I actually need it, often in non subtitled works or YouTube videos with very strange subtitle timings.

### CODE REVIEW

**Comment by tatsumoto-ren** [comment]
Hi! Before making a large PR, It would be awesome if you reached out through our community. Thanks for the changes, I'll review them.

File: helpers.lua [hunk 1]
```diff
@@ -33,6 +33,10 @@ this.max_num = function(table)
     end
     return max
 end
+this.max_n = function(note_ids, n)
+    table.sort(note_ids)
+    return {table.unpack(note_ids, math.max(#note_ids - n + 1, 1), #note_ids)}
+end
```
Comment (tatsumoto-ren) [inline helpers.lua:39]: This name is hard to understand. Apparently, this function should be named something like `take_last_n_added_notes`.

File: platform/win.lua [hunk 2]
```diff
@@ -43,8 +43,9 @@ self.curl_request = function(url, request_json, completion_fn)
         '-X',
         'POST',
         '--data-binary',
-        table.concat { '@', curl_tmpfile_path }
+        --table.concat { '@', curl_tmpfile_path }
     }
+    table.insert(args, request_json)
```
Comment (tatsumoto-ren) [inline platform/win.lua:48]: Looks like this change was made by mistake.

Comment (stexus) [reply platform/win.lua:48]: Apologies, I don't quite remember the intention of this change either. I'll remove it and verify.

**Comment by stexus** [comment]
> Hi! Before making a large PR, It would be awesome if you reached out through our community. Thanks for the changes, I'll review them.

Good to know, thanks for the heads up. I didn't see any contact in the mpvacious repo yesterday but I found the telegram on the main site today. I'll try to make an account tonight.

**Comment by tatsumoto-ren** [comment]
I can't update notes anymore using the 'm' key. After I select the subtitle lines I want, the 'm' button just doesn't do anything. 

screenshot-2024-11-26-03-44-53

File: helpers.lua [hunk 3]
```diff
@@ -33,6 +33,10 @@ this.max_num = function(table)
     end
     return max
 end
+this.max_n = function(note_ids, n)
+    table.sort(note_ids)
+    return { table.unpack(note_ids, math.max(#note_ids - n + 1, 1), #note_ids) }
```
Comment (tatsumoto-ren) [inline helpers.lua:38]: Calling`table.unpack` causes an error: `attempt to call field 'unpack' (a nil value)`. To fix the error, change `table.unpack` to `this.unpack`.

Comment (stexus) [reply helpers.lua:38]: Thank you

**Comment by stexus** [comment]
Unfortunately the multiple card update looks to be broken now. I made a few cards tonight with the regular quick card creation (with just g[0-9]), so that seems to work fine at least. Would you prefer me to close the PR and work on integration or leave it open as draft?

**Comment by stexus** [comment]
Got it, I'll add that over this week too. Currently what I have on todolist:
1. Fix multi card update functionality
2. Config support to change keybinds
3. README documentation

File: platform/win.lua (hunk 2 above)
Comment (stexus) [reply platform/win.lua]: I actually believe this was actually to prevent overwriting due to asynchronous calls? When updating multiple cards, only the last updateNoteFields goes through. I'm not 100% sure on this as I never got to the bottom of it, but I suspect it has something to do with using a temp txt file + mp.command_native_async. When I changed it to include the request_json at runtime (as the nix.lua also does), multiple card updates works again. Is there a reason for using a temporary text file for the curl commands?

**Comment by stexus** [comment]
@tatsumoto-ren Hi Tatsumoto, can you review the short tutorial I wrote in the README? 
Otherwise this is working well on my machine -- I made just shy of 50 cards so it passes the sniff test at least. Please give it a shot too! I don't use the advanced menu that much nowadays so I may have missed some corner case bugs in my testing.

File: subtitles/observer.lua [hunk 4]
```diff
@@ -194,9 +198,23 @@ self.get_timing = function(position)
     return -1
 end
 
-self.collect = function()
+self.collect = function(n_lines)
```
Comment (tatsumoto-ren) [inline subtitles/observer.lua:201]: This function does two different things depending on the argument passed. In such cases it's better to have two functions instead of one.

File: subs2srs.lua [hunk 5]
```diff
@@ -48,8 +48,9 @@ local secondary_sid = require('subtitles.secondary_sid')
 local platform = require('platform.init')
 local forvo = require('utils.forvo')
 local subs_observer = require('subtitles.observer')
-local menu
-
+local menu, quick_menu, quick_menu_card
+local n_lines
+local n_cards = 1
```
Comment (tatsumoto-ren) [inline subs2srs.lua:53]: These two values are logically tied together, so it would be appropriate to encapsulate them in one object.

**Comment by tatsumoto-ren** [comment]
Thanks for making the README. 

> Otherwise this is working well on my machine -- I made just shy of 50 cards so it passes the sniff test at least. 

It doesn't work with secondary subtitles, right? 

> Please give it a shot too! 

I'll test it.

**Comment by stexus** [comment]
> It doesn't work with secondary subtitles, right?

Yes, I didn't add support for secondary subtitles. Let me know if this is a problem.

File: subtitles/observer.lua (hunk 4 above)
Comment (stexus) [reply subtitles/observer.lua:201]: Thanks, good point.

**Comment by stexus** [comment]
Also, I saw you reverted the request_json change. Were you able to take a look at the earlier comment regarding it?

File: platform/win.lua (hunk 2 above)
Comment (tatsumoto-ren) [reply platform/win.lua]: > Is there a reason for using a temporary text file for the curl commands?

Curl on windows doesn't work reliably when json is passed as a command line argument. Using a temporary file is a widely used solution. We also ran into the issue in the past and had to adopt the fix.

* https://curl.se/mail/archive-2018-05/0011.html

>  When updating multiple cards, only the last updateNoteFields goes through. 

Then we need to use a unique filename every time.

**Comment by tatsumoto-ren** [comment]
> Yes, I didn't add support for secondary subtitles. Let me know if this is a problem.

Users will likely find this confusing.

**Comment by stexus** [comment]
Thanks, I'll give it a shot. What's the benefit of using these files for the curl command instead of passing in the request directly in the lua table?

I'll add secondary subtitle support too.

**Comment by tatsumoto-ren** [comment]
>What's the benefit of using these files for the curl command instead of passing in the request directly in the lua table?

It's to avoid the issue with failing curl requests on windows.

File: subs2srs.lua [hunk 6]
```diff
@@ -399,9 +406,16 @@ end
...
+    local n_lines = quick_creation_opts:get_lines()
+    local n_cards = quick_creation_opts:get_cards()
+    if n_lines then
+      sub = subs_observer.collect_from_all_dialogues(n_lines)
+    else
+      sub = subs_observer.collect_from_current()
+    end
     -- this now returns a table
-    local last_note_ids = ankiconnect.get_last_note_ids(n_cards or 1)
+    local last_note_ids = ankiconnect.get_last_note_ids(n_cards)
```
Comment (tatsumoto-ren) [inline subs2srs.lua]: What's the purpose of the `or 1` expression?

Comment (stexus) [reply subs2srs.lua]: I don't quite remember why I had that to be honest lol. Likely some sort of safeguard in case n_cards is nil or 0, but using the quick_creation_opts guarantees a valid value now.

**Comment by stexus** [comment]
Hi @tatsumoto-ren , sorry for the delay. I got around to testing the new curl request and secondary sub additions. Seems to be all working now. I had to modify some of the curl logic as async calls for something like grabbing the last n notes would not work.

**Comment by tatsumoto-ren** [comment]
>  I got around to testing the new curl request and secondary sub additions. Seems to be all working now. 

cool!
